saw-core/src/Verifier/SAW/Simulator/TermModel.hs

```diff
@@ -0,0 +1,1092 @@
+         reflectTerm sc cfg tp f
+
+replace ::
+  (?recordEC :: BoundECRecorder) =>
+  SharedContext ->
+  Sim.SimulatorConfig TermModel ->
+  IORef ReplaceUninterpMap ->
+  ExtCns (TValue TermModel) ->
+  IO (Value TermModel)
+replace sc cfg mapref ec = loop [] (ecType ec)
+  where
+    loop :: [Term] -> TValue TermModel -> IO (Value TermModel)
+    loop env (VPiType nm xty body) =
+        return $ VFun nm $ \x ->
+          do ty <- applyPiBody body x
+             xtm <- readBackValue sc cfg xty =<< force x
+             loop (xtm:env) ty
+
+    loop env ty =
+      do let args = reverse env
+         ty' <- readBackTValue sc cfg ty
+         newec <- scFreshEC sc (toShortName (ecName ec)) ty'
+         modifyIORef mapref (Map.alter (Just . ((newec,args):) . fromMaybe []) (ecVarIndex ec))
+         reflectTerm sc cfg ty =<< scFlatTermF sc (ExtCns newec)
+
+normalizeSharedTerm ::
+  SharedContext ->
+  ModuleMap ->
+  Map Ident TmPrim {- ^ additional primitives -} ->
+  Map VarIndex TmValue {- ^ ExtCns values -} ->
+  Set VarIndex {- ^ opaque constants -} ->
+  Term ->
+  IO Term
+normalizeSharedTerm sc m addlPrims ecVals opaqueSet t =
+  do let ?recordEC = \_ec -> return ()
+     cfg <- mfix (\cfg -> Sim.evalGlobal' m (Map.union (constMap sc cfg) addlPrims)
+                              (extcns cfg) (constants cfg) (neutral cfg) (primHandler cfg))
+     v <- Sim.evalSharedTerm cfg t
+     tv <- evalType cfg =<< scTypeOf sc t
+     readBackValue sc cfg tv v
+
+  where
+    constants cfg tf ec
+      | Set.member (ecVarIndex ec) opaqueSet = Just $
+          do let ?recordEC = \_ec -> return ()
+             tm <- scTermF sc tf
+             reflectTerm sc cfg (ecType ec) tm
+
+      | otherwise = Nothing
+
+    extcns cfg tf ec =
+      let ?recordEC = \_ec -> return () in
+      case Map.lookup (ecVarIndex ec) ecVals of
+        Just v  -> return v
+        Nothing ->
+          do tm <- scTermF sc tf
+             reflectTerm sc cfg (ecType ec) tm
+
+    neutral cfg env nt =
+      do let ?recordEC = \_ec -> return ()
+         env' <- traverse (\(x,ty) -> readBackValue sc cfg ty =<< force x) env
+         tm   <- instantiateVarList sc 0 env' =<< neutralToSharedTerm sc nt
+         tyv  <- evalType cfg =<< scTypeOf sc tm
+         reflectTerm sc cfg tyv tm
+
+    primHandler cfg pn _msg env tp =
+      do let ?recordEC = \_ec -> return ()
+         pn'  <- traverse (readBackTValue sc cfg) pn
+         args <- reverse <$> traverse (\(x,ty) -> readBackValue sc cfg ty =<< force x) env
+         prim <- scFlatTermF sc (Primitive pn')
+         f    <- foldM (scApply sc) prim args
+         reflectTerm sc cfg tp f
+
+------------------------------------------------------------
+-- Values
+
+data TermModel
+
+type TmValue = Value TermModel
+type TmPrim  = Prims.Prim TermModel
+
+type instance EvalM  TermModel = IO
+type instance VBool  TermModel = Either Term Bool
+type instance VWord  TermModel = Either (Natural, Term) BitVector
```

Comment:
It might be worth documenting what the `Natural` is here. It's perhaps easy to guess, but not entirely obvious.